python/tchannel/tornado/connection.py

```diff
@@ -190,50 +190,49 @@ def _loop(self):
             if context.message.message_type in self.CALL_TYPES:
                 self._messages.put(context)
                 continue
+
             elif context.message_id in self._outstanding:
-                message = self._response_message_factory.build(
+                response = self.response_message_factory.build(
                     context.message_id, context.message
                 )
+                if (context.message.message_type == Types.ERROR or
+                        context.message.flags == FlagsType.none):
+                    future = self._outstanding.pop(context.message_id)
+                else:
+                    # still streaming, keep it for record
+                    future = self._outstanding.get(context.message_id)
```

Comment:
Cool. I think this if-else block needs a little more explanation though. It would be very easy to break this with minor changes.